From 60279d00637e6cf1040d17c186123c8dc66081e3 Mon Sep 17 00:00:00 2001
From: ran shalit <rshalit@habana.ai>
Date: Sun, 02 May 2021 17:47:43 +0300
Subject: [PATCH] [SW-40142] gaudi: spsram: finalize rma table

Write boot info for 3 RMA tables in SP-SRAM, required
because u-boot can't access mtd in early boot, therefore
preboot does the access and pass this information to u-boot
through SP SRAM.

Change-Id: Iafdab159cfcd46d931463b9e34e22cd862ea186f
Signed-off-by: ran shalit <rshalit@habana.ai>
---

diff --git a/gaudi/rma_info.c b/gaudi/rma_info.c
index 25f31a6..119e13a 100644
--- a/gaudi/rma_info.c
+++ b/gaudi/rma_info.c
@@ -13,6 +13,7 @@
 #include "rma_if.h"
 #include "boot_if.h"
 #include "rma_info.h"
+#include "boot_info.h"
 #include "module_i2c_slave.h"
 #include "hl_i2c_slave_events.h"
 #include "i2cslave_indirect.h"
@@ -353,3 +354,29 @@
 	/* Clear out the data for buffer re-use */
 	memset(err_sts_rma_buf, 0, size);
 }
+
+#define RMA_HDR_TBL_SIZE (RMA_HEADER_ENTRY_SIZE + RMA_TABLE_HEADER_SIZE)
+
+u8 hbm_hdr_table[RMA_HDR_TBL_SIZE]__aligned(4);
+u8 hbm_ecc_table[MAX_INFO_DESC_DATA]__aligned(4);
+u8 row_rep_table[MAX_INFO_DESC_DATA]__aligned(4);
+
+void preboot_finalize_rma_tables(void)
+{
+	int ret;
+	u32 size = MAX_INFO_DESC_DATA;
+
+	ret = rma_info_get(RMA_HBM_ROW_REPLACEMENT_TABLE, row_rep_table, &size);
+	if (!ret)
+		boot_info_set(HL_BOOT_RMA_ROW_REP_TABLE, row_rep_table, size);
+
+	size = MAX_INFO_DESC_DATA;
+	ret = rma_info_get(RMA_HBM_ECC_TABLE, hbm_ecc_table, &size);
+	if (!ret)
+		boot_info_set(HL_BOOT_RMA_HBM_ECC_TABLE, hbm_ecc_table, size);
+
+	size = MAX_INFO_DESC_DATA;
+	ret = rma_info_get(RMA_HEADER_TABLE, hbm_hdr_table, &size);
+	if (!ret)
+		boot_info_set(HL_BOOT_RMA_HDR_TABLE, hbm_hdr_table, size);
+}
diff --git a/include/preboot.h b/include/preboot.h
index 8656a91..6dda31b 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -179,6 +179,7 @@
 void preboot_finalize_fw_info(void);
 void preboot_finalize_nic_en(void);
 void preboot_boot_info_set_fw_info(void);
+void preboot_finalize_rma_tables(void);
 
 uintptr_t preboot_shm_get_offset(void);
 void preboot_shm_set_offset(uintptr_t offset);
diff --git a/include/rma_info.h b/include/rma_info.h
index ba4643c..fe0e7ed 100644
--- a/include/rma_info.h
+++ b/include/rma_info.h
@@ -15,5 +15,6 @@
 void rma_info_sram_ecc_err_sts(struct hl_i2ci_error_status_sram_ecc *op);
 void rma_info_hbm_init_err_sts(struct hl_i2ci_error_status_hbm_init *op);
 void rma_info_hbm_row_replacement(struct hl_i2ci_dram *op);
+void preboot_finalize_rma_tables(void);
 
 #endif /* _RMA_INFO_H_ */
diff --git a/preboot/fit.c b/preboot/fit.c
index 4624f4b..bd11643 100644
--- a/preboot/fit.c
+++ b/preboot/fit.c
@@ -942,6 +942,7 @@
 }
 
 __weak void preboot_finalize_fw_info(void) {}
+__weak void preboot_finalize_rma_tables(void) {}
 
 void fit_run_image(void)
 {
@@ -958,6 +959,7 @@
 
 	preboot_finalize_nic_en();
 	preboot_finalize_fw_info();
+	preboot_finalize_rma_tables();
 	preboot_finalize_enabled_caps();
 	boot_info_finalize();
 	preboot_start_image(startaddr);
